les/clientpool.go

```diff
@@ -559,75 +851,87 @@ func (f *clientPool) addBalance(id enode.ID, amount int64, meta string) (uint64,
 	f.lock.Lock()
 	defer f.lock.Unlock()
 
+	now := f.clock.Now()
 	pb := f.ndb.getOrNewPB(id)
-	var negBalance uint64
+	var negBalance expiredValue
 	c := f.connectedMap[id]
 	if c != nil {
-		pb.value, negBalance = c.balanceTracker.getBalance(f.clock.Now())
+		pb.value, negBalance = c.balanceTracker.getBalance(now)
 	}
 	oldBalance := pb.value
-	if amount > 0 {
-		if amount > maxBalance || pb.value > maxBalance-uint64(amount) {
-			return oldBalance, oldBalance, errBalanceOverflow
-		}
-		pb.value += uint64(amount)
-	} else {
-		if uint64(-amount) > pb.value {
-			pb.value = 0
-		} else {
-			pb.value -= uint64(-amount)
-		}
+	posExp := f.posExpiration(now)
+	oldValue := oldBalance.value(posExp)
+	if amount > 0 && (amount > maxBalance || oldValue > maxBalance-uint64(amount)) {
+		return oldValue, oldValue, errBalanceOverflow
 	}
+	pb.value.add(amount, posExp)
```

Comment:
The expired offset is accumulated since clientpool started. So the question is: If we just create balance for a client, it seems all historical offset is stil applied.

Reply:
That is correct, the same offset is applied to all service tokens at all times. The value is multiplied by `2^(logOffset/log2Multiplier)` when it is stored and divided when converted back to actual value. This means that the decrease of value between storing and reading only depends on the relative change of `logOffset` between those two moments.
You can imagine the stored values as some "inflationary service token" whose exchange rate is always getting worse compared to the "stable service token". We use the stable version when buying, selling or spending, and the inflationary version for storing (so that they automatically expire without periodically iterating through the entire db and explicitly reducing the stored values). The exponential magic is the conversion between the two. The reason I made this weird `expiredValue` struct is that the "conversion rate" will get very high over time, possibly too big even for `float64`.

Reply:
I can understand the stored value is decreased exponentially. But I mean the new balance. Like:

- I buy 100 tokens with 100 wei
- The bought 100 tokens should be added and now my service token balance should be 100
- After x seconds, the token amount is expired gradually

However now all historical decay is also applied to new bought tokens.

- I buy 100 tokens with 100 wei
- The bought token is expired with accumulated offset, and now my balance is 35 token(e.f.)

Reply:
No, it should work as follows:
- initially, 1 "stable service token" (the amount visible through the API) equals 1 "inflationary token" (the amount stored in the database).
- you buy 100 service tokens which initially equals 100 inflationary tokens (so the number 100 is stored in the db)
- time passes, expiration happens, now 100 inflationary tokens are only worth 35 service tokens (the number 100 in the database did not change, just the `logOffset` was changed)
- now the client buys 100 more service tokens with the `buyTokens` command; this is converted to inflationary amount at the current rate, which means the database value is increased by 100 / 0.35 = 285 (there is now 385 inflationary units stored in the db)
- if the client queries its service token balance, it is calculated internally as 385*0.35 = 135 (so in the end the client sees that it has actually received 100 new service tokens as promised)